### PR TITLE
canonicalization of DO constructs

### PR DESCRIPTION
**Pull request by schweitzpgi**
Add tests and finish up DO loop canonicalization
Rewrites the parse tree

### CODE REVIEW

File: lib/semantics/canonicalize-do.cc [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+    }
+    return false;
+  }
+
+private:
+  parser::Block SpliceBlock(
+      parser::Block::iterator beginLoop, parser::Block::iterator endLoop) {
+    parser::Block block;
+    block.splice(block.begin(), *currentList_, ++beginLoop, ++endLoop);
+    return block;
```
Comment (tskeith) [inline lib/semantics/canonicalize-do.cc]: This function confuses me. It's moving elements from currentList_ to block, right? But even though block is returned, the return value isn't used. So if the goal is just to modify currentList_, isn't there a `remove` or `erase` function that does that? Even if `splice` is the best way to do this, I think the function could use a better name to make it clear what it is doing.

File: lib/semantics/canonicalize-do.cc [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+        common::Indirection<parser::DoConstruct>{std::make_tuple(
+            parser::Statement<parser::NonLabelDoStmt>{
+                std::optional<parser::Label>{},
+                parser::NonLabelDoStmt{
+                    std::make_tuple(std::optional<parser::Name>{},
+                        ExtractLoopControl(startLoop))}},
+            SpliceBlock(startLoop, endLoop),
+            parser::Statement<parser::EndDoStmt>{std::optional<parser::Label>{},
+                parser::EndDoStmt{std::optional<parser::Name>{}}})};
+    return startLoop;
```
Comment (tskeith) [inline lib/semantics/canonicalize-do.cc]: I think `std::nullopt` is better than `std::optional<...>{}` in most cases, unless you think it's important to make the type explicit.

File: lib/semantics/canonicalize-do.cc (hunk 1 above)
Comment (schweitzpgi) [reply lib/semantics/canonicalize-do.cc]: The return value is used. It is the body of the constructed DoConstruct.

File: lib/semantics/canonicalize-do.cc (hunk 2 above)
Comment (schweitzpgi) [reply lib/semantics/canonicalize-do.cc]: My compiler was whining about that, so I made them explicit to make it stop.

File: lib/semantics/canonicalize-do.cc (hunk 1 above)
Comment (tskeith) [reply lib/semantics/canonicalize-do.cc]: Thanks, I missed that. The statement started so many lines earlier that I didn't realize the call to SpliceBlock was an argument and not at statement of its own.

Comment (schweitzpgi) [reply lib/semantics/canonicalize-do.cc]: I can turn the entirety of MakeCanonicalForm() into a nested lambda expression. I don't think that would improve readability myself.

File: lib/semantics/canonicalize-do.cc [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+public:
+  CanonicalizationOfDoLoops() = default;
+
+  template<typename T> bool Pre(T &) { return true; }
+  template<typename T> void Post(T &) {}
+  bool Pre(parser::ExecutionPart &executionPart) {
+    const auto &endIter{executionPart.v.end()};
+    currentList_ = &executionPart.v;
+    for (auto iter{executionPart.v.begin()}; iter != endIter; ++iter) {
+      iter = CheckStatement(iter);
```
Comment (klausler) [inline lib/semantics/canonicalize-do.cc]: Please be sure that you want to both update `iter` with `CheckStatement` *and* increment it in the `for` statement's increment expression.  If `CheckStatement` really just checks statements, I have no idea what its returned iterator might be pointing at.

File: lib/semantics/canonicalize-do.cc [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "canonicalize-do.h"
+#include "../parser/parse-tree-visitor.h"
+
+namespace Fortran::semantics {
```
Comment (klausler) [inline lib/semantics/canonicalize-do.cc]: A lot of this code might be made more brief if it were in `namespace Fortran::parser`.  It doesn't seem to depend on much from semantics.

File: lib/semantics/canonicalize-do.cc (hunk 1 above)
Comment (klausler) [reply lib/semantics/canonicalize-do.cc]: `Splice` isn't the best name here, I agree.  Maybe `Extract`.

File: lib/semantics/canonicalize-do.cc [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+      if (auto *labelDoLoop = std::get_if<
+              parser::Statement<common::Indirection<parser::LabelDoStmt>>>(
+              &executableConstruct->u)) {
+        labelDoIters_.push_back(iter);
+        labels_.push_back(std::get<parser::Label>(labelDoLoop->statement->t));
+      } else if (!labels_.empty()) {
+        parser::Walk(executableConstruct->u, *this);
+      }
+    }
+    return currentIter_;
```
Comment (klausler) [inline lib/semantics/canonicalize-do.cc]: `currentIter_` is mysterious hidden shared state.  I don't know what it points at or which member functions can change it.

File: lib/semantics/canonicalize-do.cc [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+                        ExtractLoopControl(startLoop))}},
+            SpliceBlock(startLoop, endLoop),
+            parser::Statement<parser::EndDoStmt>{std::optional<parser::Label>{},
+                parser::EndDoStmt{std::optional<parser::Name>{}}})};
+    return startLoop;
+  }
+  parser::Block::iterator CheckStatement(const parser::Block::iterator &iter) {
+    currentIter_ = iter;
+    parser::ExecutionPartConstruct &executionPartConstruct{*iter};
+    if (auto *executableConstruct = std::get_if<parser::ExecutableConstruct>(
```
Comment (klausler) [inline lib/semantics/canonicalize-do.cc]: C++17 fixed brace initialization with `auto` and we now use braces in these situations.

File: lib/semantics/canonicalize-do.h [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef FORTRAN_SEMANTICS_CANONICALIZE_DO_H_
+#define FORTRAN_SEMANTICS_CANONICALIZE_DO_H_
+
```
Comment (klausler) [inline lib/semantics/canonicalize-do.h]: A sentence or two describing what this module does would be informative.

File: lib/semantics/canonicalize-do.cc (hunk 1 above)
Comment (tskeith) [reply lib/semantics/canonicalize-do.cc]: I don't think so either.

I think a different name for SpliceBlock would help. Maybe ExtractBlock.

File: lib/semantics/canonicalize-do.h (hunk 7 above)
Comment (schweitzpgi) [reply lib/semantics/canonicalize-do.h]: done

Comment (schweitzpgi) [reply lib/semantics/canonicalize-do.h]: done

File: lib/semantics/canonicalize-do.cc (hunk 4 above)
Comment (schweitzpgi) [reply lib/semantics/canonicalize-do.cc]: done

File: lib/semantics/canonicalize-do.cc (hunk 3 above)
Comment (schweitzpgi) [reply lib/semantics/canonicalize-do.cc]: yes, it is necessary to keep the iterator pointing at a valid list member.

File: lib/semantics/canonicalize-do.cc (hunk 1 above)
Comment (schweitzpgi) [reply lib/semantics/canonicalize-do.cc]: renamed to ExtractBlock, though the operation is std::list::splice

File: lib/semantics/canonicalize-do.cc (hunk 5 above)
Comment (schweitzpgi) [reply lib/semantics/canonicalize-do.cc]: I'll add a comment.

File: lib/semantics/canonicalize-do.cc (hunk 6 above)
Comment (schweitzpgi) [reply lib/semantics/canonicalize-do.cc]: ok

**Pull request merged**